include/CL/sycl/detail/all_true.hpp

```diff
@@ -0,0 +1,32 @@
+#ifndef TRISYCL_SYCL_DETAIL_ALL_TRUE_HPP
+#define TRISYCL_SYCL_DETAIL_ALL_TRUE_HPP
+
+/** \file Add all_true wrapper.
+    Dave Airlie
+
+    This file is distributed under the University of Illinois Open Source
+    License. See LICENSE.TXT for details.
+*/
+
+namespace cl {
+namespace sycl {
+namespace detail {
+
+template <bool...> struct bool_pack;
+
+template <bool... v>
+using all_true = std::is_same<bool_pack<true, v...>, bool_pack<v..., true>>;
```

Comment:
We need the all_true to avoid some constructor matching going wrong, otherwise I kept running into a problem where the asynchandler and the prop list would match and things got confused, so we only match in the all_true cases where everything in the prop list is a derivative of the property base class.